examples/snake/snake.cc

```diff
@@ -0,0 +1,651 @@
+		// re-drawing each frame)
+		draw_background(lcd);
+		draw_tile(lcd, snakePositions.front(), SNAKE_COLOR);
+		draw_cherry(lcd, fruitPosition);
+
+		bool gameStillActive = true;
+		while (gameStillActive)
+		{
+			uint64_t nextTime = rdcycle64();
+			uint64_t elapsedTimeMilliseconds =
+			  (nextTime - currentTime) / cyclesPerMillisecond;
+			uint64_t frameTime = MILLISECONDS_PER_FRAME;
+			if (SPEED_SCALING_ENABLED)
+			{
+				// Scale the game's speed in an inverse relationship between
+				// MILLISECONDS_PER_FRAME and MILLISECONDS_PER_FRAME / 2
+				frameTime /= 2;
+				frameTime += (frameTime / snakePositions.size());
+			}
+			if (elapsedTimeMilliseconds < frameTime)
+			{
+				uint64_t remainingTime = frameTime - elapsedTimeMilliseconds;
+				wait_with_input(remainingTime, gpio);
+			}
+			currentTime = rdcycle64();
+
+			gameStillActive = update_game_state(gpio, lcd);
+		}
+	};
+
+	/**
+	 * @brief Cleans up the non-contiguous 2D game space array used for
+	 * collision checking.
+	 */
+	void free_game_space()
+	{
+		for (size_t y = 0; y < gameSize.height; y++)
+		{
+			free(gameSpace[y]);
+		}
+		free(gameSpace);
+	}
+
+	public:
+	/**
+	 * @brief Plays a game of snake using the stored state as settings.
+	 *
+	 * @param gpio The Sonata GPIO driver to use for I/O operations.
+	 * @param lcd The LCD that will be drawn to.
+	 */
+	void run_game(volatile SonataGPIO *gpio, SonataLcd *lcd)
+	{
+		wait_for_start(gpio, lcd);
+		initialise_game();
+		main_game_loop(gpio, lcd);
+		free_game_space();
+		isFirstGame = false;
+	};
+
+	/**
+	 * @brief Constructor for a SnakeGame.
+	 *
+	 * @param lcd The LCD that the game will be drawn to.
+	 */
+	SnakeGame(SonataLcd *lcd)
+	{
+		initialise_game_size(lcd);
+	};
+
+	~SnakeGame()
+	{
+		if (gameSpace != nullptr)
+		{
+			free_game_space();
+		}
+	};
+};
+
+/**
+ * @brief A minimal function used to replace SnakeGame::check_if_colliding for
+ * use in error recovery, letting us utilise RISC-V's capability violations with
+ * a single compartment as a scoring mechanism.
+ *
+ * @return True, always.
+ */
+bool __attribute__((noinline)) returnFromHandledError()
+{
+	return true;
+}
+
+/**
+ * @brief Handles any CHERI Capability Violation errors. If the error was a
+ * Bounds or Tag violation it assumes it is because of the incorrect memory
+ * access in SnakeGame::check_if_colliding and therefore it recovers the program
+ * and ends the game. Otherwise, this force unwinds and ends the program.
+ */
+extern "C" ErrorRecoveryBehaviour
+compartment_error_handler(ErrorState *frame, size_t mcause, size_t mtval)
+{
+	auto [exceptionCode, registerNumber] = extract_cheri_mtval(mtval);
+	if (exceptionCode == CauseCode::BoundsViolation ||
+	    exceptionCode == CauseCode::TagViolation)
+	{
+		// If an explicit out of bounds access occurs, or bounds are made
+		// invalid by some negative array access, the snake has hit the boundary
+		// of the game and so the game should end.
```

Comment:
Since we don't check which part of the code this error came from. We should probably state in the comment that we *assume* the cheri violation came from the array access in `check_if_colliding`.